libraries/tlv-account-resolution/src/state.rs

```diff
@@ -158,65 +172,92 @@ impl ExtraAccountMetas {
+                }
+                Seed::AccountKey { index } => {
+                    let account_index = *index as usize;
+                    let account_meta = resolved_metas
+                        .get(account_index)
+                        .ok_or::<ProgramError>(AccountResolutionError::AccountNotFound.into())?;
+                    pda_seeds.push(account_meta.pubkey.as_ref());
+                }
+            }
         }
-        Ok(())
+        Ok(Pubkey::find_program_address(&pda_seeds, program_id).0)
     }
 
     /// Add the additional account metas to an existing instruction
     pub fn add_to_instruction<T: SplDiscriminate>(
         instruction: &mut Instruction,
         data: &[u8],
     ) -> Result<(), ProgramError> {
-        Self::add_to_vec::<T>(&mut instruction.accounts, data)
+        let state = TlvStateBorrowed::unpack(data)?;
+        let bytes = state.get_first_bytes::<T>()?;
+        let extra_account_metas = PodSlice::<PodAccountMeta>::unpack(bytes)?;
+
+        for extra_meta in extra_account_metas.data().iter() {
+            let mut account_meta = match extra_meta.discriminator {
+                0 => AccountMeta::try_from(extra_meta)?,
+                1 => {
+                    let seeds = Seed::unpack_array(&extra_meta.address_config)?;
+                    AccountMeta {
+                        pubkey: Self::resolve_pda(
+                            &instruction.accounts,
+                            &instruction.data,
+                            &instruction.program_id,
+                            &seeds,
+                        )?,
+                        is_signer: extra_meta.is_signer.into(),
+                        is_writable: extra_meta.is_writable.into(),
+                    }
+                }
+                _ => return Err(ProgramError::InvalidAccountData),
+            };
+            Self::de_escalate_account_meta(&mut account_meta, &instruction.accounts);
+            instruction.accounts.push(account_meta);
+        }
+
+        Ok(())
     }
 
-    /// Add the additional account metas and account infos for a CPI, while
-    /// de-escalating repeated accounts.
-    ///
-    /// If an added account already exists in the instruction with lower
-    /// privileges, match it to the existing account. This prevents a lower
-    /// program from gaining unexpected privileges.
+    /// Add the additional account metas and account infos for a CPI
     pub fn add_to_cpi_instruction<'a, T: SplDiscriminate>(
         cpi_instruction: &mut Instruction,
         cpi_account_infos: &mut Vec<AccountInfo<'a>>,
         data: &[u8],
         account_infos: &[AccountInfo<'a>],
     ) -> Result<(), ProgramError> {
-        let state = TlvStateBorrowed::unpack(data)?;
-        let bytes = state.get_first_bytes::<T>()?;
-        let extra_account_metas = PodSlice::<PodAccountMeta>::unpack(bytes)?;
+        let initial_instruction_metas = cpi_instruction.accounts.clone();
```

Comment:
nit: Rather than cloning the vector, can we just take the initial length and search within that subslice?